### PR TITLE
(3/5) Migrate SourceSwitchHandler to use source modules

### PR DESCRIPTION
**Pull request by jkonecny12**
This have to be merged together with others PRs otherwise it won't work.

This is based on commits from previous PRs.

### CODE REVIEW

File: pyanaconda/ui/helpers.py [hunk 1]
```diff
@@ -94,7 +95,10 @@ def unset_source(self):
         release all related resources (unmount iso files, drop caches, etc.).
         """
         self._clean_hdd_iso()
-        self.data.method.method = None
+
+        payload_proxy = get_payload()
+        payload_proxy.SetSources([])
+
         payloadMgr.restart_thread(self.payload, checkmount=False)   # pylint: disable=no-member
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:102]: I think we could try to remove the method `unset_source`.

File: pyanaconda/ui/helpers.py [hunk 2]
```diff
@@ -104,7 +108,14 @@ def _clean_hdd_iso(self):
         This means unmounting the partition and unprotecting it,
         so it can be used for the installation.
         """
-        if self.data.method.method == "harddrive" and self.data.method.partition:
+        payload_proxy = get_payload()
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:111]: Please, use `self.payload`, `self.payload.proxy`, `self.payload.get_source_proxy` and `set_source`.

File: pyanaconda/ui/helpers.py [hunk 3]
```diff
@@ -104,7 +108,14 @@ def _clean_hdd_iso(self):
...
+
+        try:
+            source_proxy = get_source(payload_proxy)
+        except ValueError:
+            return
+
+        if source_proxy.Type == constants.SOURCE_TYPE_HDD and source_proxy.Partition:
             unmark_protected_device(self.data.method.partition)
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:119]: You should replace the partition here as well.

File: pyanaconda/ui/helpers.py [hunk 4]
```diff
@@ -115,67 +126,79 @@ def set_source_hdd_iso(self, device_name, iso_path):
...
-        # the / gets stripped off by payload.ISO_image
-        self.data.method.dir = "/" + iso_path
-
-        # as we already made the device_name protected when
-        # switching to it, we don't need to protect it here
+        payload_proxy = get_payload()
+        payload_proxy.SetSources([get_object_path(new_source_proxy)])
 
-    def set_source_url(self, url=None):
+    def set_source_url(self, url):
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:142]: Please, see changes required by GUI and TUI in the class SourceSwitchHandler (https://github.com/rhinstaller/anaconda/pull/2526).

File: pyanaconda/ui/helpers.py [hunk 5]
```diff
@@ -115,67 +126,79 @@ def set_source_hdd_iso(self, device_name, iso_path):
...
         """ Switch to install source specified by URL """
         # clean any old HDD ISO sources
         self._clean_hdd_iso()
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:145]: I would rename the method so something like `_clean_sources`. We need to call the tear-down task before we set up the sources.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Implemented notes from @poncovka . Thanks!

File: pyanaconda/ui/helpers.py [hunk 6]
```diff
@@ -115,67 +102,75 @@ def set_source_hdd_iso(self, device_name, iso_path):
...
-        self._clean_hdd_iso()
+        self._tear_down_existing_source()
+
+        url_source_proxy = create_source(constants.SOURCE_TYPE_URL)
+
+        repo_conf = RepoConfigurationData()
+        repo_conf.url = url
+        repo_conf.type = url_type
+        if proxy:
+            repo_conf.proxy = proxy
```
Comment (poncovka) [inline pyanaconda/ui/helpers.py:128]: You can use `proxy or ""`.

Comment (jkonecny12) [reply pyanaconda/ui/helpers.py:128]: Good improvement.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Implemented tweak suggested by @poncovka

File: pyanaconda/ui/helpers.py [hunk 7]
```diff
@@ -76,36 +75,24 @@ class SourceSwitchHandler(object, metaclass=ABCMeta):
...
-        payloadMgr.restart_thread(self.payload, checkmount=False)   # pylint: disable=no-member
-        threadMgr.wait(constants.THREAD_PAYLOAD_RESTART)
-        threadMgr.wait(constants.THREAD_PAYLOAD)
-
-    def _clean_hdd_iso(self):
+    def _tear_down_existing_source(self):
         """ Clean HDD ISO usage
         This means unmounting the partition and unprotecting it,
         so it can be used for the installation.
         """
```
Comment (VladimirSlavik) [inline pyanaconda/ui/helpers.py:89]: The docstring no longer matches, I think. The cleaning feels like side effect...

Comment (jkonecny12) [reply pyanaconda/ui/helpers.py:89]: True, thanks!

**Comment by jkonecny12** [comment]
**UPDATED:**
* Fix suggested by @VladimirSlavik . Thanks!

**Pull request merged**